File: app/src/main/java/ca/appvelopers/mcgillmobile/ui/settings/SettingsActivity.java
/*
 * Copyright 2014-2017 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ca.appvelopers.mcgillmobile.ui.settings;

import android.content.Context;
import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.content.FileProvider;
import android.widget.LinearLayout;

import com.guerinet.formgenerator.FormGenerator;
import com.guerinet.utils.Utils;
import com.guerinet.utils.dialog.DialogUtils;
import com.guerinet.utils.prefs.BooleanPreference;
import com.raizlabs.android.dbflow.sql.language.SQLite;

import java.io.File;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

import javax.inject.Inject;
import javax.inject.Named;

import butterknife.BindView;
import butterknife.ButterKnife;
import ca.appvelopers.mcgillmobile.App;
import ca.appvelopers.mcgillmobile.BuildConfig;
import ca.appvelopers.mcgillmobile.R;
import ca.appvelopers.mcgillmobile.model.AppUpdate;
import ca.appvelopers.mcgillmobile.ui.DrawerActivity;
import ca.appvelopers.mcgillmobile.ui.dialog.list.HomepagesAdapter;
import ca.appvelopers.mcgillmobile.util.dagger.prefs.PrefsModule;
import ca.appvelopers.mcgillmobile.util.dagger.prefs.UsernamePreference;
import ca.appvelopers.mcgillmobile.util.manager.HomepageManager;
import okio.BufferedSink;
import okio.Okio;
import timber.log.Timber;

/**
 * Allows the user to change the app settings
 * @author Julien Guerinet
 * @since 1.0.0
 */
public class SettingsActivity extends DrawerActivity {
    /**
     * The {@link FormGenerator} container
     */
    @BindView(R.id.container)
    LinearLayout container;
    /**
     * Statistics BooleanPreference
     */
    @Inject
    @Named(PrefsModule.STATS)
    BooleanPreference statsPref;
    /**
     * 24 hour time BooleanPreference
     */
    @Inject
    @Named(PrefsModule.SCHEDULE_24HR)
    BooleanPreference twentyFourHourPref;
    /**
     * {@link UsernamePreference} instance
     */
    @Inject
    UsernamePreference usernamePref;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings);
        ButterKnife.bind(this);
        App.component(this).inject(this);
        setTitle(getString(R.string.settings_version, Utils.versionName(this)));
        analytics.sendScreen("Settings");

        FormGenerator fg = FormGenerator.bind(container);

        // 24 hour time preference
        fg.aSwitch()
                .text(R.string.settings_twentyfourhours)
                .leftIcon(R.drawable.ic_clock)
                .checked(twentyFourHourPref.get())
                .onCheckChanged((buttonView, isChecked) -> twentyFourHourPref.set(isChecked))
                .build();

        // Homepage choice
        fg.text()
                .text(homepageManager.getTitleString())
                .leftIcon(R.drawable.ic_phone_android)
                .onClick(item -> DialogUtils.list(this, R.string.settings_homepage_title,
                        new HomepagesAdapter(this) {
                            @Override
                            public void onHomepageSelected(@HomepageManager.Homepage int choice) {
                                // Update the instance
                                homepageManager.set(choice);

                                analytics.sendEvent("Settings", "HomepageManager",
                                        homepageManager.getString());

                                // Update the TextView
                                item.view().setText(homepageManager.getTitleString());
                            }
                        }))
                .build();

        // Statistics
        fg.aSwitch()
                .text(R.string.settings_statistics)
                .leftIcon(R.drawable.ic_trending_up)
                .checked(statsPref.get())
                .onCheckChanged((buttonView, isChecked) -> statsPref.set(isChecked))
                .build();

        // Help
        fg.text()
                .text(R.string.title_help)
                .leftIcon(R.drawable.ic_help)
                .onClick(item -> startActivity(new Intent(this, HelpActivity.class)))
                .build();

        // About
        fg.text()
                .text(R.string.title_about)
                .leftIcon(R.drawable.ic_info)
                .onClick(item -> startActivity(new Intent(this, AboutActivity.class)))
                .build();

        // Bug Report
        fg.text()
                .text(R.string.title_report_bug)
                .leftIcon(R.drawable.ic_bug_report)
                .onClick(item -> {
                    analytics.sendEvent("About", "Report a Bug");

                    Intent intent = new Intent(Intent.ACTION_SEND);

                    // Recipient
                    intent.putExtra(Intent.EXTRA_EMAIL, new String[]
                            {"julien.guerinet@mail.mcgill.ca"});

                    // Title
                    intent.putExtra(Intent.EXTRA_SUBJECT, getString(R.string.bug_report));

                    // Content
                    String device = "Device: " + Utils.device();
                    String sdkVersion = "SDK Version: " + Build.VERSION.SDK_INT;
                    String appVersion = "App Version: " + Utils.versionName(this);
                    String language = "Language: " + Locale.getDefault().getLanguage();

                    ConnectivityManager manager = (ConnectivityManager)
                            getSystemService(Context.CONNECTIVITY_SERVICE);
                    NetworkInfo info = manager.getActiveNetworkInfo();

                    String connection = "Connection Type: N/A";
                    if (info != null) {
                        connection = "Connection Type: " + info.getTypeName() + " " +
                                info.getSubtypeName();
                    }

                    String content = "===============" +
                            "\nDebug Info" +
                            "\n===============" +
                            "\n" + device +
                            "\n" + sdkVersion +
                            "\n" + appVersion +
                            "\n" + language +
                            "\n" + connection +
                            "\n===============\n\n";
                    intent.putExtra(Intent.EXTRA_TEXT, content);

                    // Log everything before printing the logs so it's included
                    Timber.i(device);
                    Timber.i(sdkVersion);
                    Timber.i(appVersion);
                    Timber.i(language);
                    Timber.i(connection);

                    ArrayList<Uri> uriList = new ArrayList<>();

                    // Logs (attachment)
                    try {
                        File file = new File(getExternalFilesDir(null), "logs.txt");
                        Utils.getLogs(file);
                        uriList.add(FileProvider.getUriForFile(this, BuildConfig.APPLICATION_ID +
                                ".fileProvider", file));
                    } catch (IOException e) {
                        Timber.e(new Exception("Error getting logs", e));
                    }

                    // Update logs (attachment)
                    try {
                        List<AppUpdate> appUpdates = SQLite.select()
                                .from(AppUpdate.class)
                                .queryList();

                        // Create the file that will hold the update logs
                        File file = new File(getExternalFilesDir(null), "update_logs.log");

                        // Create the Okio bugger to write the logs
                        BufferedSink sink = Okio.buffer(Okio.sink(file));

                        for (AppUpdate update : appUpdates) {
                            // Write the updates to the file
                            sink.writeUtf8("Version: " + update.getVersion() + "\nDate: " +
                                    update.getTimestamp().toString() + "\n\n");
                        }
                        sink.flush();

                        uriList.add(FileProvider.getUriForFile(this, BuildConfig.APPLICATION_ID +
                                ".fileProvider", file));
                    } catch (Exception e) {
                        Timber.e(e, "Error attaching update logs to feedback/bug report email");
                    }

                    intent.putParcelableArrayListExtra(Intent.EXTRA_STREAM, uriList);

                    // Code (Email)
                    intent.setType("message/rfc822");
                    startActivity(Intent.createChooser(intent, null));
                })
                .build();
    }

    @HomepageManager.Homepage
    @Override
    protected int getCurrentPage() {
        return HomepageManager.SETTINGS;
    }
}

File: app/src/main/java/ui/settings/SettingsActivity.kt
/*
 * Copyright 2014-2019 Julien Guerinet
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.guerinet.mymartlet.ui.settings

import android.content.Intent
import android.os.Bundle
import android.util.Pair
import com.guerinet.morf.TextViewItem
import com.guerinet.morf.morf
import com.guerinet.morf.util.Position
import com.guerinet.mymartlet.BuildConfig
import com.guerinet.mymartlet.R
import com.guerinet.mymartlet.ui.DrawerActivity
import com.guerinet.mymartlet.ui.settings.about.AboutActivity
import com.guerinet.mymartlet.util.Prefs
import com.guerinet.mymartlet.util.manager.HomepageManager
import com.guerinet.room.UpdateDao
import com.guerinet.suitcase.coroutines.bgDispatcher
import com.guerinet.suitcase.coroutines.uiDispatcher
import com.guerinet.suitcase.dialog.singleListDialog
import com.guerinet.suitcase.io.getFileUri
import com.guerinet.suitcase.log.TimberTag
import com.guerinet.suitcase.prefs.BooleanPref
import com.guerinet.suitcase.util.Utils
import com.guerinet.suitcase.util.extensions.openUrl
import kotlinx.android.synthetic.main.activity_settings.*
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import okio.buffer
import okio.sink
import org.jetbrains.anko.startActivity
import org.koin.android.ext.android.inject
import java.io.File
import java.io.IOException
import java.util.Comparator

/**
 * Allows the user to change the app settings
 * @author Julien Guerinet
 * @since 1.0.0
 */
class SettingsActivity : DrawerActivity(), TimberTag {

    override val tag: String = "SettingsActivity"

    override val currentPage = HomepageManager.HomePage.SETTINGS

    private val twentyFourHourPref by inject<BooleanPref>(Prefs.SCHEDULE_24HR)

    private val updateDao by inject<UpdateDao>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings)
        title = getString(R.string.settings_version, BuildConfig.VERSION_NAME)

        container.morf {

            // 24 hour time preference
            aSwitch {
                textId = R.string.settings_twentyfourhours
                icon(Position.START, R.drawable.ic_clock)
                isChecked = twentyFourHourPref.value
                onCheckChanged { _, isChecked ->
                    twentyFourHourPref.value = isChecked
                }
            }

            // Home Page
            text {
                text = homePageManager.titleString
                icon(Position.START, R.drawable.ic_phone_android)
                onClick { item: TextViewItem ->
                    // Get the list of home pages in alphabetical order for the current language
                    val homePages = HomepageManager.HomePage.values()
                        .map { Pair(it, homePageManager.getTitle(it)) }
                        .sortedWith(Comparator { o1, o2 -> o1.second.compareTo(o2.second) })

                    val currentChoice =
                        homePages.indexOfFirst { it.first == homePageManager.homePage }

                    singleListDialog(homePages.map { it.second }, R.string.settings_homepage_title, currentChoice) {
                        // Update the instance
                        homePageManager.homePage = homePages[it].first

                        // Update the TextView
                        item.text = homePageManager.titleString

                        analytics.event("settings", "homepage" to homePageManager.title)
                    }
                }
            }

            // Help
            text {
                textId = R.string.title_help
                icon(Position.START, R.drawable.ic_help)
                onClick { startActivity<HelpActivity>() }
            }

            // About
            text {
                textId = R.string.title_about
                icon(Position.START, R.drawable.ic_info)
                onClick { startActivity<AboutActivity>() }
            }

            // Privacy Policy
            text {
                text = "Privacy Policy"
                icon(Position.START, R.drawable.ic_info)
                onClick {
                    openUrl("https://github.com/jguerinet/MyMartlet/blob/master/privacy-policy.md")
                }
            }

            // Bug Report
            text {
                textId = R.string.title_report_bug
                icon(Position.START, R.drawable.ic_bug_report)
                onClick {
                    analytics.event("report_bug")

                    val intent = Intent(Intent.ACTION_SEND).apply {

                        // Code (Email)
                        type = "message/rfc822"

                        // Recipient
                        putExtra(Intent.EXTRA_EMAIL, arrayOf("julien.guerinet@mail.mcgill.ca"))

                        // Title
                        putExtra(Intent.EXTRA_SUBJECT, getString(R.string.bug_report))

                        // Content
                        val content =
                            Utils.getDebugInfo(BuildConfig.VERSION_NAME, BuildConfig.VERSION_CODE)

                        // Log everything in the logs
                        timber.i(content)

                        putExtra(Intent.EXTRA_TEXT, content)
                    }

                    launch(uiDispatcher) {

                        withContext(bgDispatcher) {
                            // Update logs (attachment)
                            try {
                                val appUpdates = updateDao.getAll()

                                // Create the file that will hold the update logs
                                val file = File(getExternalFilesDir(null), "update_logs.log")

                                // Create the Okio buffer to write the logs
                                val sink = file.sink().buffer()

                                appUpdates.forEach { update ->
                                    // Write the updates to the file
                                    sink.writeUtf8(
                                        "Version: ${update.version}\nDate: ${update.timestamp}\n\n"
                                    )
                                }
                                sink.flush()

                                intent.putExtra(
                                    Intent.EXTRA_STREAM,
                                    getFileUri(BuildConfig.APPLICATION_ID, file)
                                )
                            } catch (e: IOException) {
                                timber.e(e, "Error attaching update logs to bug report email")
                            }
                        }

                        startActivity(Intent.createChooser(intent, null))
                    }
                }
            }
        }
    }
}
